scipy/_lib/_util.py

```diff
@@ -1111,11 +1111,38 @@ def _dict_formatter(d, n=0, mplus=1, sorter=None):
 "core" shapes. However, array argument(s) of this function may have additional
 "batch" dimensions prepended to the core shape. In this case, the array is treated
 as a batch of lower-dimensional slices; see :ref:`linalg_batch` for details.
-Note that calls with zero-size batches are unsupported and will raise a ``ValueError``.
 """
 
 
-def _apply_over_batch(*argdefs):
+def output_from_signature(arrays, batch_shape, core_shapes, signature):
+    dtype = np.result_type(*arrays)
+
+    # parse more efficiently with regex?
+    inputs, outputs = signature.split("->")
+    inputs, outputs = inputs[1:-1].split("),("), outputs[1:-1].split("),(")
+    input_dim_to_letter = {}
+    for i, input in enumerate(inputs):
+        for j, l in enumerate(input.split(",")):
+            input_dim_to_letter[(i, j)] = l
+
+    letter_to_length = {}
+    for i, core_shape in enumerate(core_shapes):
+        for j, length in enumerate(core_shape):
+            l = input_dim_to_letter[(i, j)]
+            if hasattr(letter_to_length, l):
+                assert letter_to_length[l] == length
```

Comment:
This is just to demonstrate the sort of input validation that *could* be done automatically if we add NEP 5 signature information. In particular, this checks that if (say) `i` appears in the signature in multiple place, then the corresponding elements of the input shapes are all consistent.